src/main/scala/uk/gov/nationalarchives/omega/jms/JmsRRClient.scala

```diff
@@ -0,0 +1,204 @@
+  }
+
+  /**
+   * Create a JMS Request-Reply Client for use with Amazon Simple Queue Service.
+   *
+   * @param credentials the credentials for connecting to the ActiveMQ broker.
+   * @param customClientId an optional Custom client ID to identify this client.
+   *
+   * @param replyQueue the queue that replies should be consumed from.
+   *
+   * @return The resource for the JMS Request-Reply Client.
+   */
+  def createForSqs[F[_]: Async: Logger](endpoint: HostBrokerEndpoint, credentials: UsernamePasswordCredentials, customClientId: Option[F[String]] = None)(replyQueue: String): Resource[F, JmsRRClient[F]] = {
+    val clientIdRes: Resource[F, String] = Resource.liftK[F](customClientId.getOrElse(RandomClientIdGen.randomClientId[F]))
+
+    val jmsClientRes: Resource[F, JmsClient[F]] = clientIdRes.flatMap { clientId =>
+      simpleQueueService.makeJmsClient[F](simpleQueueService.Config(
+        endpoint = simpleQueueService.Endpoint(Some(DirectAddress(HTTP,endpoint.host,Some(endpoint.port))),"elasticmq"),
+        credentials = Some(Credentials(credentials.username, credentials.password)),
+        clientId = simpleQueueService.ClientId(clientId),
+        None
+      ))
+    }
+
+    create[F](jmsClientRes)(replyQueue)
+  }
+
+  /**
+   * Create a JMS Request-Reply Client.
+   *
+   * @param jmsClientRes a jms4s Client resource.
+   *
+   * @param replyQueue the queue that replies should be consumed from.
+   *
+   * @return The resource for the JMS Request-Reply Client.
+   */
+  def create[F[_]: Async: Sync: Logger](jmsClientRes: Resource[F, JmsClient[F]])(replyQueue: String): Resource[F, JmsRRClient[F]] = {
+    for {
+      requestMap <- Resource.pure(new ConcurrentHashMap[String, ReplyMessage => F[Unit]](16, 0.75f, defaultProducerConcurrencyLevel))
+      jmsClient <- jmsClientRes
+      consumer <- jmsClient.createAcknowledgerConsumer(QueueName(replyQueue), concurrencyLevel = defaultConsumerConcurrencyLevel, pollingInterval = defaultConsumerPollingInterval)
+      consumerHandlerRes = consumer.handle(jmsConsumerHandler[F](requestMap)(_, _)).background
+      producerRes = jmsClient.createProducer(concurrencyLevel = defaultProducerConcurrencyLevel)
+
+      // tie the life-times of consumerHandler and producer together
+      consumerProducer <- Resource.both(consumerHandlerRes, producerRes)
+    } yield new JmsRRClient[F](requestMap)(consumerProducer._1, consumerProducer._2)
+  }
+
+  /**
+   * A jms4s consumer handler that consumes a reply message,
+   * finds the ReplyMessageHandler and dispatches the message
+   * to it.
+   */
+  private def jmsConsumerHandler[F[_]: Async: Logger](requestMap: ConcurrentHashMap[String, ReplyMessageHandler[F]])(jmsMessage: JmsMessage, mf: MessageFactory[F])(implicit F: Async[F], L: Logger[F]): F[AckAction[F]] = {
+    val maybeCorrelatedRequestHandler: F[Option[ReplyMessageHandler[F]]] = F.delay(jmsMessage.getJMSCorrelationId.flatMap(correlationId => Option(requestMap.remove(correlationId))))
```

Comment:
Sometimes I can see an entry in the request map, however the correlation ID does not match the correlation ID from the JMS message.

Reply:
I'm afraid I can't reproduce this now: https://github.com/nationalarchives/jms4s-request-reply/pull/1#discussion_r1093207104

Reply:
What happens if the frontend sends a message to the service, then crashes? Would the message not get lost in that case, as the frontend no longer has a reference to this correlation ID?